libs/log/sem_logger.go

```diff
@@ -0,0 +1,266 @@
+	l.semRulesLock.Lock()
+	defer l.semRulesLock.Unlock()
+
+	_, exist := l.rules[entity]
+	key := string(value)
+	if !exist {
+		l.rules[entity] = map[string]uint{key: 0}
+	} else {
+		_, exist = l.rules[entity][key]
+		if !exist {
+			l.rules[entity][key] = 0
+		}
+	}
+	l.setEnabled(true)
+}
+
+func (l *semLogger) DeleteAll() {
+	func() {
+		l.semRulesLock.Lock()
+		defer l.semRulesLock.Unlock()
+
+		for rule := range l.rules {
+			delete(l.rules, rule)
+		}
+	}()
+
+	l.setEnabled(false)
+	l.setActiveState(false)
+}
+
+func (l *semLogger) Status() (status SemStatus, err error) {
+	status = SemStatus{
+		Active: l.isActive(),
+		Rules:  map[SemEntity]map[string]bool{},
+	}
+
+	l.semRulesLock.RLock()
+	defer l.semRulesLock.RUnlock()
+	for entity, rules := range l.rules {
+		for rule, count := range rules {
+			_, exists := status.Rules[entity]
+			if !exists {
+				status.Rules[entity] = map[string]bool{}
+			}
+			status.Rules[entity][rule] = count > 0
+
+		}
+	}
+	return
+}
+
+// Set SEM logger state to active
+// In active state all log levels are enabled.
+// In deactivated state (no matching rule) only error logs are forwarded to logger backend
+func (l *semLogger) setActiveState(active bool) {
+	// For Filter loggers we need to make sure that no filtering is happening underneath
+	next, isFilterLogger := l.logBackend.(*filter)
+	if isFilterLogger {
+		//Filter backend needs to bypass all logs incase of SEM is enabled
+		next.SetSemStatus(!l.isDisabled())
+	}
+	l.active = active
+}
+
+// Disable/Enable SEM logger
+// In disabled state (no rules defined) no 'filterring' is done on SEM level and no bypass
+// on filter-capable backends should be done
+func (l *semLogger) setEnabled(enable bool) {
+	l.enabled = enable
+	// For Filter loggers we need to make sure that no filtering is happening underneath
+	next, isFilterLogger := l.logBackend.(*filter)
+	if isFilterLogger {
+		next.SetSemStatus(enable)
+	}
+}
+
+func (l *semLogger) Entry(entity SemEntity, value []byte) {
+	if !l.ruleExists(entity, value) {
+		return
+	}
+	if !l.active {
+		l.setActiveState(true)
+	}
+	l.ruleCountIncrease(entity, value)
+}
+
+func (l *semLogger) ruleExists(entity SemEntity, rule []byte) bool {
+	l.semRulesLock.RLock()
+	defer l.semRulesLock.RUnlock()
+	_, exist := l.rules[entity][string(rule)]
+	return exist
+}
+
+func (l *semLogger) ruleGetCount(entity SemEntity, rule []byte) (uint, bool) {
+	l.semRulesLock.RLock()
+	defer l.semRulesLock.RUnlock()
+	count, exist := l.rules[entity][string(rule)]
+	return count, exist
+
+}
+
+func (l *semLogger) ruleCountIncrease(entity SemEntity, rule []byte) uint {
+	if !l.ruleExists(entity, rule) {
+		return 0
+	}
+
+	l.semRulesLock.Lock()
+	defer l.semRulesLock.Unlock()
+	l.rules[entity][string(rule)]++
+	return l.rules[entity][string(rule)]
+}
+
+func (l *semLogger) ruleCountDecrease(entity SemEntity, rule []byte) uint {
+	if !l.ruleExists(entity, rule) {
+		return 0
+	}
+
```

Comment:
What if another goroutine runs "delete all" here?
I have the impression we need to lock/unlock only when entering/exiting the public methods, to avoid these corner cases.

Reply:
it would be blocked until there are no active read or write locks anymore.
As long as there is no notify mechanism or central SEM state we need the low-level Rlock on the forwarding at least.